service/src/main/kotlin/fi/espoo/evaka/varda/VardaUpdateServiceV2.kt

```diff
@@ -0,0 +1,698 @@
+            handleUpdatedEvakaServiceNeed(db, client, serviceNeedId)
+        } else {
+            handleNewEvakaServiceNeed(db, client, serviceNeedId)
+        }
+    }
+}
+
+fun handleDeletedEvakaServiceNeed(db: Database.Connection, client: VardaClient, serviceNeedId: UUID) {
+    try {
+        val vardaServiceNeed = db.read { it.getVardaServiceNeed(serviceNeedId) }
+        if (vardaServiceNeed != null) {
+            val errors = deleteServiceNeedDataFromVarda(client, vardaServiceNeed)
+            if (errors.isNotEmpty()) {
+                logger.error(errors.joinToString(","))
+                db.transaction { it.markServiceNeedUpdateFailed(serviceNeedId, errors) }
+            } else {
+                db.transaction { it.deleteVardaServiceNeedByEvakaServiceNeed(serviceNeedId) }
+            }
+        }
+    } catch (e: Exception) {
+        logger.error("VardaUpdate: manual check needed: something went from while trying to delete varda service need $serviceNeedId data: ${e.localizedMessage}")
+    }
+}
+
+fun handleUpdatedEvakaServiceNeed(db: Database.Connection, client: VardaClient, serviceNeedId: UUID) {
+    try {
+        val vardaServiceNeed = db.read { it.getVardaServiceNeed(serviceNeedId) } ?: throw Exception("VardaUpdate: cannot update service need $serviceNeedId: varda service need missing")
+        val deleteErrors = deleteServiceNeedDataFromVarda(client, vardaServiceNeed)
+        if (deleteErrors.isNotEmpty()) {
+            logger.error(deleteErrors.joinToString(","))
+            db.transaction {
+                it.markServiceNeedUpdateFailed(serviceNeedId, deleteErrors)
+            }
+        } else {
+            val evakaServiceNeed = db.read { it.getEvakaServiceNeedInfoForVarda(serviceNeedId) }
+            val addErrors = addServiceNeedDataToVarda(db, client, evakaServiceNeed, vardaServiceNeed)
+            if (addErrors.isNotEmpty()) {
+                logger.error(addErrors.joinToString(","))
+                db.transaction {
+                    it.markServiceNeedUpdateFailed(serviceNeedId, addErrors)
+                }
+            }
+        }
+        db.transaction { it.upsertVardaServiceNeed(vardaServiceNeed) }
+    } catch (e: Exception) {
+        logger.error("VardaUpdate: manual check needed: something went from while trying to add varda service need $serviceNeedId data: ${e.localizedMessage}")
+    }
+}
+
+fun handleNewEvakaServiceNeed(db: Database.Connection, client: VardaClient, serviceNeedId: UUID) {
+    try {
+        val evakaServiceNeed = db.read { it.getEvakaServiceNeedInfoForVarda(serviceNeedId) }
+        val newVardaServiceNeed = evakaServiceNeedToVardaServiceNeed(evakaServiceNeed.childId, evakaServiceNeed)
+        db.transaction { it.upsertVardaServiceNeed(newVardaServiceNeed) }
+
+        val errors = addServiceNeedDataToVarda(db, client, evakaServiceNeed, newVardaServiceNeed)
+        db.transaction { it.upsertVardaServiceNeed(newVardaServiceNeed) }
```

Comment:
Also just to make sure... Is it intentional that this is run twice here? Or is it just my partial understanding that doesn't grasp any obvious rationale for it.

Reply:
This was for error handling: first we create the varda service need baseline data on (on line 131), then all collected information on 134. If for some reason the line 133 would throw an exception, without line 131 we would lose the information. I refactored the addServiceNeedDataToVarda on Friday so that it basically cannot throw exception anymore, so yes, line 131 is unnecessary and now remoted.